### ISSUE TITLE
ToggleClass has update issues when using a sorted filtered list.

### ISSUE THREAD

**Issue by snorhax** [edvin/tornadofx#1068]
I am using a set of toggle class functions that highlight cells in table rows according to the row item's due date and status:

```
tableview(tasks) {
             column("Due Date", TaskData::dueDateProperty) {
                 cellFormat {
                      text = it.format(dateFormat)
                      val today = LocalDate.now()
                      toggleClass(Style.good, it.isAfter(today) && rowItem.status < 2)
                      toggleClass(Style.okay, it.isEqual(today) && rowItem.status < 2)
                      toggleClass(Style.bad, it.isBefore(today) && rowItem.status < 2)
                 }
             }
.
.
.
```

The table uses a sorted filtered list for it's data, however, and when the predicate for the list is changed, unrelated rows are also being affected by the toggle class functions. I'm guessing this is because they are old recycled rows that didn't have their style class reset. I've also noticed that if I shrink the window holding the table view and then unshrink it, the table redraws properly with no styling issues. Here's a few images of the issue after toggling the checkbox a few times.
![image](https://user-images.githubusercontent.com/11637807/64362564-5c4bd200-cfd4-11e9-9e47-51828d601905.png)
![image2](https://user-images.githubusercontent.com/11637807/64362577-61a91c80-cfd4-11e9-85b6-c0a37002d220.png)
Is there a way to force the table view to redraw its cells or am I not using the toggleClass function correctly?

**Comment by edvin** [comment]
This approach is no good, since the classes will be bound to the table cell even when it's not showing data anymore. You should instead use `TableCellFragment` and add the css classes to a ui element inside the view, which in turn should be bound to an ItemViewModel inside the fragment.

**Comment by snorhax** [comment]
Wasn't sure what you meant but I found enough bits and pieces through various posts to make this instead. Probably still isn't proper but it works:
```
cellFactory = Callback {
                        object : TableCell<TaskData, LocalDate>() {
                            override fun updateItem(item: LocalDate?, empty: Boolean) {
                                if (empty || item == null) {
                                    text = null
                                    graphic = null
                                    removeClass(Style.good, Style.okay, Style.bad)
                                } else {
                                    text = item.toString()
                                    if(rowItem.status < 2) {
                                        val today = LocalDate.now()
                                        when {
                                            item.isAfter(today) -> addClass(Style.good)
                                            item.isEqual(today) -> addClass(Style.okay)
                                            item.isBefore(today) -> addClass(Style.bad)
                                            else -> removeClass(Style.good, Style.okay, Style.bad)
                                        }
                                    }
                                }
                            }
                        }
                    }
```

**Comment by snorhax** [comment]
Just a note, I eventually did make a similar `ListCellFragment` with an `ItemViewModel` that tried to bind with a toggle class. There were still similar problems, so I had to abandon the toggle class and have the fragment change the CSS class by listening to both the specific property in the model and listening to when the itemProperty of the fragment changed. Basically, it seems that toggle class isn't very helpful if the rows are due to change.

**Comment by edvin** [comment]
ListCellFragment should handle this just fine, provided you have an ItemViewModel and bind against it state. If you have a complete runnable sample I can try to figure out what's going wrong. Please keep the sample as small as possible with only relevant pieces of code :)

**Comment by snorhax** [comment]
I ended up preparing an example for you but in the process I think I found my issue:
`val model = MyModel().bindTo(this)`
It seemed that style classes would reset because models would rollback upon not being visibly shown anymore (due to scrolling, shrinking window, etc.). However, this can be fixed if the model is committed (of course the commit logic is up to the developer). I can only guess that this was the problem earlier as I don't really have that much time to dwell on it, but I hope this helps anyone else who's new to getting their hands dirty with these in-depth concepts.
